features/cellular/easy_cellular/EasyCellularConnection.h

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef EASY_CELLULAR_CONNECTION_H
+#define EASY_CELLULAR_CONNECTION_H
+
+#include "CellularConnectionUtil.h"
+#ifdef CELLULAR_DEVICE
```

Comment:
Having done this, you should be able to cross-link to "OnboardCellularInterface.h" so that it includes this and typedefs `OnboardCellularInterface` as this object (ifdef CELLULAR_DEVICE), which should(?) totally eliminate the need for any app changes.

The intent was always that people providing onboard modems should arrange for them to appear as `OnboardCellularInterface`.

Reply:
If application has defined CELLULAR_DEVICE (in mbed_app.json or mbed target for onboard modem) then modem is selected from cellular/targets/ folder. If application wants to use OnboardCellularInterface.h then it needs to specifically instantiate some class implementing that. That choice has now been intentionally left to application developers.

Reply:
When designing the last round of cellular, what we ended up with was an attempt to achieve the "best of both worlds" for modem selection, while aligning with general design mbed OS architectural style.

For Ethernet, applications can usually just say `EthernetInterface`, because the target arranges for that to "do the right thing", accessing the Ethernet usually built into the main chip. For an add-on Wi-fi board, the application has to say `XXXWiFi(pin, pin, pin)`.

The style in mbed OS so far has been for "on-main-chip" devices to be part of the "target", and hence automatically available to portable applications without explicit specification. You just specify the target. Plug-in things have to be explicitly specified, so applications doing that are not automatically portable.

Part of what I've been trying to work toward generally, and discussing with @-sg,  is to allow the same driver (or a variant) to be both available for manual specification, and to be "just there" if it is actually built in to a target. The bulk of review (and redesign) last time was on this point.

The intent with OnboardCellularInterface was that someone could have a custom `XXXModem(pin, pin, pin)` that would work flexibly, and on any target. But if they actually built it into a target, the target would arrange for OnboardCellularInterface to be based on that general driver, with default constructor using all the correct pins and power controls, so that the application gets the same ease of use they get for EthernetInterface, due to it being target-specified.

Your framework would fit into that mechanism. If your target.json is defining CELLULAR_DEVICE to arrange for the right driver to be present because it's on the board, it makes sense for you to be pointing OnboardCellularInterface at that.

If you were talking about a non-target-specified modem, then the general style would be for the application to specify the modem by invoking its specific class, rather than JSON/define setting. I don't think applications should be setting CELLULAR_DEVICE (or its equivalent) - they should be giving it by name. That does mean you wouldn't have an equivalent to "easy cellular" for the offboard, but my understanding that that was out of immediate scope, and we were assuming on-board for now.

I'll spend some time discussing this with @-sg today or tomorrow to try to make sure we've got common architectural agreement here, then we can agree what to do on this.

Reply:
Oh, for reference and comparison with other non-cellular things, which are works in progress, see: https://github.com/ARMmbed/mbed-os/pull/6108

There you see "OnboardModemInterface" as fitting in as one of the sub-defaults of the overall "default network interface".

It's not following the new pattern I'm establishing - that would be `CellularBase::get_default_instance()` - but it achieves a similar effect. Maybe it could be deprecated in favour of that sort of call?

Reply:
Right, the decision is that we will be doing the `CellularBase::get_default_instance()` and its the equivalents for non-cellular for 5.9, but for 5.8 we just stick with the existing 5.7 abstraction design - use `OnboardCellularInterface` as the abstraction point.

Now that you have something which should be a fully-fledged replacement for the 4 in-tree targets defining `MODEM_ON_BOARD_UART` (Ublox C027 & C030 and the 2 Dragonfly boards) including that APN functionality, I think you can switch those to the new code by changing the structure of OnboardCellularInterface to flow something like this:

    #include "EasyCellularConnection.h"
    #ifdef CELLULAR_DEVICE
      typedef EasyCellularConnection OnboardCellularInterface;
    #elif MODEM_ON_BOARD && MODEM_ON_BOARD_UART && NSAPI_PPP_AVAILABLE
      #include "UARTCellularInterface.h"
      class OnboardCellularInterface : public UARTCellularInterface;
    #endif

That should mean that the client and cellular examples work without modification on those 4 targets and your new ones, as long as CELLULAR_DEVICE is being set automatically by the FEATURE_TARGET filtering. (Are there any remaining issues about necessary defines/configs?)  

Any other targets we don't know about which set MODEM_ON_BOARD_UART will continue to use the 5.7 PPP driver. (In principle though, can't you're framework have a generic/default driver that would be able to act like UARTCellularInterface - I still think the Lisa U one is basically generic? Something to think about when removing that functionality).

In 5.9, we will have `CellularBase::get_default_instance()`, and that will give you more flexibility than that simple typedef - you would be able to construct something and maybe do more configuration. The OnboardCellularInterface point would be deprecated in favour of that.

We can also do `CellularDevice::get_default_instance()` for the full API then.

Annoyingly I see one error in mbed-os-example-cellular - it's calling `iface.modem_debug_on`, and that's not a CellularBase method, so would fail. Docs do say    

    Depending on the type of on-board modem, OnboardCellularInterface
    could be derived from different implementation classes.
    Portable applications should only rely on it being a CellularBase.

Adding that method to EasyCellular would avoid that compile failing, or we could remove the call from the app or make it compile time conditional somehow.

Reply:
Typedef onboard cellular to easy cellular:
https://github.com/AriParkkila/mbed-os/commit/a3d878328a0f5adfb43d17d4ba88efee59692504

Reply:
Now after typedef https://github.com/ARMmbed/mbed-os-example-cellular can use this PR without any changes.